presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUtil.java

```diff
@@ -746,4 +764,120 @@ public static void loadCachingProperties(Map<String, String> properties, Iceberg
+            }
+        };
+    }
+
+    private static class DeleteFilesIterator
+            implements CloseableIterator<DeleteFile>
+    {
+        private final Set<String> seenFiles = new HashSet<>();
+        private final Map<Integer, PartitionSpec> partitionSpecsById;
+        private final CloseableIterator<FileScanTask> fileTasks;
+        private final Optional<Set<Integer>> requestedPartitionSpec;
+        private final Optional<Set<Integer>> requestedSchema;
+        private Iterator<DeleteFile> currentDeletes = emptyIterator();
+        private DeleteFile currentFile;
+
+        private DeleteFilesIterator(Map<Integer, PartitionSpec> partitionSpecsById,
+                CloseableIterator<FileScanTask> fileTasks,
+                Optional<Set<Integer>> requestedPartitionSpec,
+                Optional<Set<Integer>> requestedSchema)
+        {
+            this.partitionSpecsById = partitionSpecsById;
+            this.fileTasks = fileTasks;
+            this.requestedPartitionSpec = requestedPartitionSpec;
+            this.requestedSchema = requestedSchema;
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return currentFile != null || advance();
+        }
+
+        private boolean advance()
+        {
+            currentFile = null;
+            while (currentFile == null && (currentDeletes.hasNext() || fileTasks.hasNext())) {
+                if (!currentDeletes.hasNext()) {
+                    currentDeletes = fileTasks.next().deletes().iterator();
+                }
+                while (currentDeletes.hasNext()) {
+                    DeleteFile deleteFile = currentDeletes.next();
+                    if (shouldIncludeFile(deleteFile)) {
+                        // If there is a requested schema only include files that match it
+                        if (seenFiles.add(deleteFile.path().toString())) {
+                            currentFile = deleteFile;
+                            return true;
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public DeleteFile next()
+        {
+            DeleteFile result = currentFile;
+            advance();
+            return result;
+        }
+
+        private boolean shouldIncludeFile(DeleteFile file)
+        {
+            boolean matchesPartition = !requestedPartitionSpec.isPresent() ||
+                    requestedPartitionSpec.get().equals(partitionSpecsById.get(file.specId()).fields().stream().map(PartitionField::fieldId).collect(Collectors.toSet()));
+            return matchesPartition && (file.content() == FileContent.POSITION_DELETES ||
+                    !requestedSchema.isPresent() || requestedSchema.get().equals(ImmutableSet.copyOf(file.equalityFieldIds())));
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            fileTasks.close();
```

Comment:
I believe this might need to be explicitly set to `emptyIterator` as well, similar to https://github.com/prestodb/presto/pull/21605/commits/18f4226603ceb06659500254e4e3c1e9a71c4f86#r1442237365

Reply:
Done 👍